stdlib/public/Concurrency/Actor.cpp

```diff
@@ -929,9 +939,14 @@ class DefaultActorImpl : public HeapObject {
   // the future
   alignas(sizeof(ActiveActorStatus)) char StatusStorage[sizeof(ActiveActorStatus)];
 
-  using ListMerger = swift::ListMerger<Job *, JobQueueTraits>;
-  ListMerger::LastInsertionPoint lastInsertionPoint =
-      ListMerger::LastInsertionPoint();
+  // All jobs are stored in a single linked list
+  // Head of the list is stored inside ActiveActorStatus
+  // List starts with a zero or more unprocessed jobs of mixed priorities in LIFO order.
+  // After that list contains processed jobs sorted by priority.
+  // Within the same priority jobs are sorted in the FIFO order.
+  // Elements of `insertionPoints` point to the last job of the corresponding priority,
+  // or are null if no jobs of that priority exist.
+  Job *insertionPoints[NumPriorityBuckets];
```

Comment:
Hmm.  This isn't what I was thinking, but it probably still works.  I was imagining that preprocessing would break down the incoming queue into entirely separate linked lists.  That would let us eliminate all of the `needsPreprocessing` logic: the expectation would be that nothing would stay in the incoming queue after preprocessing.   On the other hand, we'd need some way to efficiently find the end of each priority list.  `job->SchedulerPrivate` has space for a doubly-linked list (which would let us make priority escalation also O(1)), or we could just store the tail pointer separately if we have space for two pointers per priority bucket.

Let me know which you'd prefer to implement — if you're inclined to try out the separate queues approach, I'll save myself the trouble of reviewing the patch in its current state.

Reply:
>  Or we could just store the tail pointer separately if we have space for two pointers per priority bucket.
 
I've started with this idea, but then observed that we really need to store only separate tails. Storing separate heads  needlessly uses limited real estate in the `DefaultActorImpl`, and also requires to add logic of "if this list is empty, then check the next one". Joining the lists together solves the latter, and saves space as a bonus.

I don't think that storing heads, and then following the back pointer would be more efficient that storing tail pointers directly. And maintaining two links during insertion and removal will only increase costs.

Even if storing heads separately and enqueueing job to the right list, I'm afraid we would still need preprocessing logic. AFAIU, entire preprocessing logic comes from the fact it is pretty straightforward to atomically insert at the beginning at list, while atomically inserting at the end is quite challenging. Probably not impossible, but I'd need to do some research before I am able to implement it. AFAIU, the main goal of the preprocessing is to reverse the list.

We probably could separate head of processed jobs (in FIFO order) from the head of unprocessed jobs (in LIFO order). But I am not sure if this brings any benefits.